exercises/building_suffix_arrays/descr.md

```diff
@@ -0,0 +1,543 @@
+0 (included) and 10 (excluded):
+
+```
+  mississipi  0
+   ississipi  1
+    ssissipi  2
+     sissipi  3
+      issipi  4
+       ssipi  5
+        sipi  6
+         ipi  7
+          pi  8
+           i  9
+```
+
+Sorting this list lexicographically produces the following result:
+
+```
+  i           9
+  ipi         7
+  issipi      4
+  ississipi   1
+  mississipi  0
+  pi          8
+  sipi        6
+  sissipi     3
+  ssipi       5
+  ssissipi    2
+```
+
+Thus, the suffix array associated with the string `mississipi` is
+this array of integers:
+
+```
+              9
+              7
+              4
+              1
+              0
+              8
+              6
+              3
+              5
+              2
+```
+
+that is, in OCaml notation, the array `[|9; 7; 4; 1; 0; 8; 6; 3; 5; 2|]`.
+
+## Preliminaries
+
+In this exercise, we manipulate many arrays. Thus, we wish to refer to
+the most common operations on arrays via short names. Our prelude makes
+the following definitions:
+
+```
+  let copy = Array.copy
+  let init = Array.init
+  let iter = Array.iter
+  let length = Array.length
+  let make = Array.make
+  let map = Array.map
+```
+
+These functions are provided by the module
+[Array](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Array.html)
+of the OCaml standard library.
+We also use several of the functions provided by the modules
+[String](https://caml.inria.fr/pub/docs/manual-ocaml/libref/String.html)
+and
+[List](https://caml.inria.fr/pub/docs/manual-ocaml/libref/List.html),
+but do not introduce short names for these functions.
+
+As announced earlier, a suffix of a string `s` is represented as an index into
+the string `s`, its start index. Thus, we make the following type definition:
+
+```
+  type suffix = int
+```
+
+While programming and debugging, it is convenient to have a function that
+converts a value of type `suffix` to an actual suffix, that is, a string.
+By mapping this function over a suffix array, we obtain a function that
+converts a suffix array to a human-readable form. Thus, our prelude makes
+the following definitions:
+
+```
+  let suffix (s : string) (i : suffix) : string =
+    let n = String.length s in
+    String.sub s i (n-i)
+
+  let suffixes (s : string) (a : suffix array) : string array =
+    map (suffix s) a
+```
+
+As an example, try evaluating the OCaml expression
+`suffixes "mississipi" [|9; 7; 4; 1; 0; 8; 6; 3; 5; 2|]`
+and check that you get the result that you expect.
+
+Because the function `suffixes` turns a suffix array of size `n` into a string
+array whose size is proportional to `n^2`, it should never be used as part of
+an efficient algorithm. It is meant to be used only while programming and
+debugging.
+
+Our prelude also defines a function that extracts a prefix out of a string:
+
+```
+  let prefix (h : int) (s : string) : string =
+    let n = String.length s in
+    if n <= h then s else String.sub s 0 h
+```
+
```

Comment:
You should probably explain why you have this special case for `h` greater than the length of `s`. The fact that you lose `String.length (prefix h s) = h` disturbs me a bit.

Reply:
(Sorry for the late reply.) I don't think I have a special case. The function call `String.sub s 0 h` is permitted only when `h <= n` holds, and the law `String.length (prefix h s) = h` holds only when `h <= n` holds, too. So, I am just extending the function `prefix` so that it works also when `h > n`. That said, this extension is probably useless, so we could replace it with `assert (h <= n)`, if you prefer.